src/session.c

```diff
@@ -0,0 +1,463 @@
+}
+
+static void
+close_sessions_in_thread_func (GTask *task,
+                               gpointer source_object,
+                               gpointer task_data,
+                               GCancellable *cancellable)
+{
+  const char *sender = (const char *)task_data;
+  GSList *list = NULL;
+  GSList *l;
+  GHashTableIter iter;
+  Session *session;
+
+  G_LOCK (sessions);
+  if (sessions)
+    {
+      g_hash_table_iter_init (&iter, sessions);
+      while (g_hash_table_iter_next (&iter, NULL, (gpointer *)&session))
+        {
+          if (strcmp (sender, session->sender) == 0)
+            list = g_slist_prepend (list, g_object_ref (session));
+        }
+    }
+  G_UNLOCK (sessions);
+
+  for (l = list; l; l = l->next)
+    {
+      Session *session = l->data;
+
+      SESSION_AUTOLOCK (session);
+      session_close (session, FALSE);
+    }
+
+  g_slist_free_full (list, g_object_unref);
+}
+
+void
+close_sessions_for_sender (const char *sender)
+{
+  g_autoptr(GTask) task = NULL;
+
+  task = g_task_new (NULL, NULL, NULL, NULL);
+  g_task_set_task_data (task, g_strdup (sender), g_free);
+  g_task_run_in_thread (task, close_sessions_in_thread_func);
+}
+
+static void
+on_closed (XdpImplSession *object)
+{
+  Session *session = (Session *)object;
+
+  SESSION_AUTOLOCK_UNREF (g_object_ref (session));
+
+  g_clear_object (&session->impl_session);
+  session_close (session, TRUE);
+}
+
+static gboolean
+session_initable_init (GInitable *initable,
+                       GCancellable *cancellable,
+                       GError **error)
+{
+  Session *session = (Session *)initable;
+  g_autofree char *sender_escaped = NULL;
+  g_autofree char *id = NULL;
+  g_autoptr(XdpImplSession) impl_session = NULL;
+  int i;
+
+  sender_escaped = g_strdup (session->sender + 1);
+  for (i = 0; sender_escaped[i]; i++)
+    {
+      if (sender_escaped[i] == '.')
+        sender_escaped[i] = '_';
+    }
+
+  if (!session->token)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED, "Missing token");
+      return FALSE;
+    }
+
+  id = g_strdup_printf ("/org/freedesktop/portal/desktop/session/%s/%s",
+                        sender_escaped, session->token);
+
+  impl_session = xdp_impl_session_proxy_new_sync (session->impl_connection,
+                                                  G_DBUS_PROXY_FLAGS_NONE,
+                                                  session->impl_dbus_name,
+                                                  id,
+                                                  NULL, error);
+  if (!impl_session)
+    return FALSE;
+
+  g_signal_connect (impl_session, "closed", G_CALLBACK (on_closed), NULL);
+
+  session->id = g_steal_pointer (&id);
+  session->impl_session = g_steal_pointer (&impl_session);
+
+  g_dbus_interface_skeleton_set_flags (G_DBUS_INTERFACE_SKELETON (session),
+                                       G_DBUS_INTERFACE_SKELETON_FLAGS_HANDLE_METHOD_INVOCATIONS_IN_THREAD);
+
```

Comment:
I think you should add a g-authorize-method callback here that only allows callers where the sender matches the sender part of the object path